src/ILCompiler.Compiler/src/Compiler/DependencyAnalysis/DictionaryLayoutNode.cs

```diff
@@ -0,0 +1,158 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+
+using Internal.TypeSystem;
+
+using ILCompiler.DependencyAnalysisFramework;
+
+namespace ILCompiler.DependencyAnalysis
+{
+    /// <summary>
+    /// Represents the layout of the generic dictionary associated with a given canonical
+    /// generic type or generic method. Maintains a bag of <see cref="DictionaryEntry"/> associated
+    /// with the canonical entity.
+    /// </summary>
+    /// <remarks>
+    /// The generic dictionary doesn't have any dependent nodes because <see cref="DictionaryEntry"/>
+    /// are runtime-determined - the concrete dependency depends on the generic context the canonical
+    /// entity is instantiated with.
+    /// </remarks>
+    class DictionaryLayoutNode : DependencyNodeCore<NodeFactory>
+    {
+        private TypeSystemEntity _owningMethodOrType;
+        private HashSet<DictionaryEntry> _entries = new HashSet<DictionaryEntry>();
+        private DictionaryEntry[] _layout;
+
+        public DictionaryLayoutNode(TypeSystemEntity owningMethodOrType)
+        {
+            _owningMethodOrType = owningMethodOrType;
+            Validate();
+        }
+
+        public void EnsureEntry(DictionaryEntry entry)
+        {
+            // TODO: thread safety
```

Comment:
Fixed. I didn't put much effort into coming up with a way to assert that nobody is adding entries while we're computing layout, but that shouldn't happen.